pkg/networkpolicy/dnsagent.go

```diff
@@ -0,0 +1,252 @@
+	logLevel int
+	cache    *ipCache
+}
+
+func (n *DNSCacheAgent) Run(ctx context.Context) error {
+	logger := klog.FromContext(ctx)
+
+	err := n.syncRules()
+	if err != nil {
+		return err
+	}
+	config := nflog.Config{
+		Group:    nfLogGroup,
+		Copymode: nflog.CopyPacket,
+		Bufsize:  4096,
+	}
+
+	nf, err := nflog.Open(&config)
+	if err != nil {
+		return fmt.Errorf("could not open nflog socket: %v", err)
+	}
+	defer nf.Close()
+
+	// Avoid receiving ENOBUFS errors.
+	if err := nf.SetOption(netlink.NoENOBUFS, true); err != nil {
+		return fmt.Errorf("failed to set netlink option %v: %v",
+			netlink.NoENOBUFS, err)
+	}
+
+	// hook that is called for every received packet by the nflog group
+	hook := func(attrs nflog.Attribute) int {
+		packet, err := parsePacket(*attrs.Payload)
+		if err != nil {
+			logger.Error(err, "Can not process packet")
+			return 0
+		}
+		// Just print out the payload of the nflog packet
+		logger.V(n.logLevel).Info("Evaluating packet", "packet", packet)
+		return 0
+	}
+
+	// errFunc that is called for every error on the registered hook
+	errFunc := func(e error) int {
+		// Just log the error and return 0 to continue receiving packets
+		klog.Infof("received error on hook: %v", e)
+		return 0
+	}
+
+	// Register your function to listen on nflog group 100
+	err = nf.RegisterWithErrorFunc(ctx, hook, errFunc)
+	if err != nil {
+		return fmt.Errorf("failed to register hook function: %v", err)
+	}
+
+	// Block till the context expires
+	<-ctx.Done()
+	return nil
+}
+
+func (n *DNSCacheAgent) syncRules() error {
+	klog.V(2).Info("Syncing kube-network-policies dnscache nftables rules")
+	nft, err := nftables.New()
+	if err != nil {
+		return fmt.Errorf("fastpath failure, can not start nftables:%v", err)
+	}
+
+	// add + delete + add for flushing all the table
+	table := &nftables.Table{
+		Name:   tableName,
+		Family: nftables.TableFamilyINet,
+	}
+	nft.AddTable(table)
+	nft.DelTable(table)
+	nft.AddTable(table)
+
+	chain := nft.AddChain(&nftables.Chain{
+		Name:     "postrouting",
+		Table:    table,
+		Type:     nftables.ChainTypeFilter,
+		Hooknum:  nftables.ChainHookPostrouting,
+		Priority: nftables.ChainPriorityLast,
+	})
+	logFlags := expr.LogFlagsIPOpt + expr.LogFlagsTCPOpt + expr.LogFlagsUID
+	snaplen := uint32(maxDNSSize)
+	// Log UDP DNS answers
+	// TODO(tcp)
+	/*
+	  [ meta load l4proto => reg 1 ]
+	  [ cmp eq reg 1 0x00000011 ]
+	  [ payload load 2b @ transport header + 0 => reg 1 ]
+	  [ cmp eq reg 1 0x00003500 ]
+	  [ log group 100 snaplen 0 qthreshold 0 ]
+	*/
+	nft.AddRule(&nftables.Rule{
+		Table: table,
+		Chain: chain,
+		Exprs: []expr.Any{
+			&expr.Meta{Key: expr.MetaKeyL4PROTO, SourceRegister: false, Register: 0x1},
+			&expr.Cmp{Op: 0x0, Register: 0x1, Data: []byte{syscall.IPPROTO_UDP}},
+			&expr.Payload{DestRegister: 0x1, Base: expr.PayloadBaseTransportHeader, Offset: 0, Len: 2},
+			&expr.Cmp{Op: expr.CmpOpEq, Register: 0x1, Data: binaryutil.BigEndian.PutUint16(53)},
```

Comment:
Given the chain definition, this rule runs on the post-DNAT packet, right? You can't assume port 53 then. (OpenShift uses `targetPort: 5353` on the CoreDNS pods so they don't need `CAP_NET_BIND_SERVICE`.)

The equivalent of `ct original proto-dst` should work.

Reply:
it is the postrouting on the return packet, so is the answer the pod will receive ... my thinking is

pod a --- > dns port 53

we capture just before it gets delivered to the pod, so it has to have src port 53 at that time